[needs revision] Use su if sudo is not available

devnsec-com opened the pull request:
This PR fixes #1148

devnsec-com commented:
Tested on (`sudo` uninstalled) CentOS 7 with command line arguments.

devnsec-com commented:
I think the travis CI is doing weird thing again:
```
Traceback (most recent call last):

  File "/home/travis/build/letsencrypt/letsencrypt/.tox/py27/bin/letsencrypt", line 9, in <module>

    load_entry_point('letsencrypt==0.1.0.dev0', 'console_scripts', 'letsencrypt')()

  File "/home/travis/build/letsencrypt/letsencrypt/letsencrypt/cli.py", line 1133, in main

    return args.func(args, config, plugins)

  File "/home/travis/build/letsencrypt/letsencrypt/letsencrypt/cli.py", line 479, in obtaincert

    le_client = _init_le_client(args, config, authenticator, installer)

  File "/home/travis/build/letsencrypt/letsencrypt/letsencrypt/cli.py", line 174, in _init_le_client

    acc, acme = _determine_account(args, config)

  File "/home/travis/build/letsencrypt/letsencrypt/letsencrypt/cli.py", line 161, in _determine_account

    config, account_storage, tos_cb=_tos_cb)

  File "/home/travis/build/letsencrypt/letsencrypt/letsencrypt/client.py", line 87, in register

    acme = _acme_from_config_key(config, key)

  File "/home/travis/build/letsencrypt/letsencrypt/letsencrypt/client.py", line 35, in _acme_from_config_key

    verify_ssl=(not config.no_verify_ssl))

  File "/home/travis/build/letsencrypt/letsencrypt/acme/acme/client.py", line 60, in __init__

    self.net.get(directory).json())

  File "/home/travis/build/letsencrypt/letsencrypt/acme/acme/client.py", line 599, in get

    self._send_request('GET', url, **kwargs), content_type=content_type)

  File "/home/travis/build/letsencrypt/letsencrypt/acme/acme/client.py", line 581, in _send_request

    response = requests.request(method, url, *args, **kwargs)

  File "/home/travis/build/letsencrypt/letsencrypt/.tox/py27/lib/python2.7/site-packages/requests/api.py", line 50, in request

    response = session.request(method=method, url=url, **kwargs)

  File "/home/travis/build/letsencrypt/letsencrypt/.tox/py27/lib/python2.7/site-packages/requests/sessions.py", line 468, in request

    resp = self.send(prep, **send_kwargs)

  File "/home/travis/build/letsencrypt/letsencrypt/.tox/py27/lib/python2.7/site-packages/requests/sessions.py", line 576, in send

    r = adapter.send(request, **kwargs)

  File "/home/travis/build/letsencrypt/letsencrypt/.tox/py27/lib/python2.7/site-packages/requests/adapters.py", line 423, in send

    raise ConnectionError(e, request=request)

ConnectionError: HTTPConnectionPool(host='localhost', port=4000): Max retries exceeded with url: /directory (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x7fa5e0063cd0>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

I'm closing and re-opening this PR to trigger it again, hopefully can resolve this issue.

kuba commented on letsencrypt-auto:
```diff
@@ -14,7 +14,15 @@ VENV_PATH=${VENV_PATH:-"$XDG_DATA_HOME/$VENV_NAME"}
+    for i in "${!args[@]}"; do
+      args[$i]="'$(printf "%s" "${args[$i]}" | sed -e "s/'/'\"'\"'/g")' "
+    done
+    exec su root -c "$0 ${args[*]}"
```
The same problem as in https://github.com/letsencrypt/letsencrypt/pull/1191#discussion-diff-43545036. We don't want entire script to be run as root.

devnsec-com replied:
If the script does not run as root, it will require the user enter their password many times.

TheNavigat replied:
https://github.com/letsencrypt/letsencrypt/pull/1191#issuecomment-152668800

kuba commented:
If this is happening, then Travis is affected by boulder#756 (https://github.com/letsencrypt/letsencrypt/wiki/Known-issues#developers).

kuba commented:
This PR duplicates efforts from #1191. Can you coordinate guys?

devnsec-com commented:
> If this is happening, then Travis is affected by boulder#756 (https://github.com/letsencrypt/letsencrypt/wiki/Known-issues#developers).

Now, again. Close and reopen this PR.....

TheNavigat commented:
Great work, actually, however, you missed something.

The parentheses of the function will cause an error if you use ```sh``` and not ```bash```, which is mentioned here. https://unix.stackexchange.com/questions/45781/shell-script-fails-syntax-error-unexpected

Otherwise, good work, although I'm not sure if something else is wrong.

devnsec-com commented:
@TheNavigat Thanks for point this out, it was my fault, I forgot this part when I wrote that code.

I have also found another bug, which is the customized `sudo` function will always replace `sudo` command, because this is shell script, not Python script.

Now, this version should be fully compatible to `dash`, `bash`, `zsh`, etc. (But does not include `csh`, since it is not a POSIX compliant shell)

pde commented on letsencrypt-auto:
```diff
@@ -14,7 +14,19 @@ VENV_PATH=${VENV_PATH:-"$XDG_DATA_HOME/$VENV_NAME"}
 
 if test "`id -u`" -ne "0" ; then
-  SUDO=sudo
+  if type sudo 1>/dev/null 2>&1; then
```
`type` doesn't seem to be part of the [POSIX shell standard](http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html), though it's supported in a lot of shells.

pde replied:
`command -v` actually is a [POSIX semi-standard](http://pubs.opengroup.org/onlinepubs/009696899/utilities/command.html). 

I used `which` for python versoin-wrangling elsewhere in the shell; we should probably just use one of these...

pde replied:
`command -v` actually is a [POSIX semi-standard](http://pubs.opengroup.org/onlinepubs/009696899/utilities/command.html). 

I used `which` for python versoin-wrangling elsewhere in the shell; we should probably just use one of these...

pde commented on letsencrypt-auto:
```diff
@@ -14,7 +14,19 @@ VENV_PATH=${VENV_PATH:-"$XDG_DATA_HOME/$VENV_NAME"}
+    su_sudo() {
+      args=""
+      while [ $# -ne 0 ]; do
+        args="$args'$(printf "%s" "$1" | sed -e "s/'/'\"'\"'/g")' "
```
I'm amazed this works, though it appears to do so across a few different shells and execution cases that I've tested.  It's probably robust enough for this use case (which is just the $SUDO'd commmands in this file), though it might be good to add a comment explaining it a little: why do `'` quotes need special treatment when other special quote types (<tt>` " \ </tt>) don't seem to?

pde commented:
This isn't strictly a `1.0 for launch` ticket but if we keep getting PRs about it maybe we should merge this anyway.

devnsec-com commented:
I have added the comments for the `su_sudo()` function, make it easier to understand for everyone.
I also changed `type` command to `command -v` as @pde suggested.

pde commented:
Sorry, I should have merged this sooner.  Better late than never!

pde commented:
@devnsec-com Thanks for some hairy-but-well-documened shell hacking!
